### ISSUE TITLE
Compressed bash-4.4 having problems running xmalloc

### ISSUE THREAD

**Issue by romeojulietthotel** [upx/upx#5]
These are some I ran into while building ffmpeg:

/bin/sh: xmalloc: variables.c:459: cannot allocate 1813 bytes (40960 bytes allocated)
/bin/sh: xmalloc: hashlib.c:57: cannot allocate 4096 bytes (32768 bytes allocated)
/bin/sh: xmalloc: setlinebuf.c:45: cannot allocate 1008 bytes (0 bytes allocated)
/bin/sh: xmalloc: hashlib.c:57: cannot allocate 4096 bytes (24576 bytes allocated)
/bin/sh: xmalloc: hashlib.c:57: cannot allocate 4096 bytes (32768 bytes allocated)
/bin/sh: xmalloc: setlinebuf.c:45: cannot allocate 1008 bytes (0 bytes allocated)
/bin/sh: xmalloc: hashlib.c:57: cannot allocate 4096 bytes (24576 bytes allocated)
/bin/sh: xmalloc: hashlib.c:57: cannot allocate 1024 bytes (57344 bytes allocated)
/bin/sh: xmalloc: hashlib.c:57: cannot allocate 4096 bytes (32768 bytes allocated)

This is all linux x86_64  and /bin/sh is a link to /bin/bash.

**Comment by jreiser** [comment]
Please specify the exact version number (echo $BASH_VERSION), the packaging version number from your Linux distro, and "sha1sum /bin/sh".  [You know, a real bug report: enough to reproduce the problem that you see.]

**Comment by romeojulietthotel** [comment]
I did give you most of the info. Here it is in discrete form.

1. download bash-4.4 tar ball (just released)
2. build bash
3. strip bash
4. upx --ultra-brute bash
5. use this new bash 
6. build ffmpeg ( I built from git )

That's what I did. Might be able to hit this building something else that's smaller. Maybe even rebuilding bash itself.

Oh, also you might find inspiration in how https://github.com/rg3/youtube-dl/ handles new issue reports.

**Comment by markus-oberhumer** [comment]
Works for me.

**Comment by romeojulietthotel** [comment]
Found a simpler test case:

 ```
./ld-linux-x86-64.so.2 /bin/bash.upx -c 'echo 123'
Segmentation fault

./ld-linux-x86-64.so.2 /bin/bash -c 'echo 123'    
123
```

I even tried changing randomize_va_space but that made no difference.

**Comment by jreiser** [comment]
```
$ strip bash; upx -o bash.upx bash; ls -l bash.upx
-rwxrwxr-x. 1 user group  491712 Oct  2 15:02 bash-upxh
$ readelf --headers bash.upx
Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  LOAD           0x0000000000000000 0x0000000000400000 0x0000000000400000
                 0x0000000000077dc4 0x0000000000077dc4  R E    200000
  LOAD           0x0000000000110a20 0x0000000000710a20 0x0000000000710a20
                 0x0000000000000000 0x0000000000000000  RW     200000
```
Now consider the second [PT_]LOAD of bash.upx, whose sole purpose is to set the value of brk(0), which is 0x710a20 (calculated as (.p_memsz + .p_vaddr) from the second PT_LOAD of the original bash.)  Because the .p_align is 0x200000 (2MiB), then the low 21 bits of .p_offset must agree with the low 21 bits of 0x710a20; so the .p_offset must be 0x110a20 which is 1116704, and this value is greater than 491712 which is the total size of bash.upx.  Explicitly running ld-linux diagnoses this condition:
```
$ /lib64/ld-linux-x86-64.so.2 ./bash-upx -c 'echo 123'
./bash-upx: error while loading shared libraries: ./bash-upx: ELF load command past end of file
```
Note that the second PT_LOAD has 0 ==  .p_filesz == .p_memsz.   ld-linux chooses to complain about the specific representation of that empty region, while the linux kernel knows better (isomorphic empty regions of length 0 occur at EVERY virtual address) and does not care.

Thus upx has done "too good a job": the result of compressing the stripped bash is too small be represented if you are as picky as ld-linux.  If that bothers you, then pad the output with some zero bytes.  (Actually, any bytes will work.)  Or, don't strip bash before compressing.  Then upx will compress and save the "extra" bytes (symbol tables, etc. that are not mmap()ed by execve()) so that "upx -d" can recover the entire original file, and the total size of the result will be more than the value of (brk(0) % .p_align).

**Comment by jreiser** [comment]
Commit e3fe0dc89b1f7b04135647ed5a0fc9e5b5e1d683 avoids ld-linux complaining about (filesize < .p_offset).

After that, running under explicit ld-linux triggers a SIGSEGV bug in ld-linux.  See https://bugzilla.redhat.com/show_bug.cgi?id=1381315 "SIGSEGV when executable lacks PT_DYNAMIC"

**Comment by romeojulietthotel** [comment]
IIUC there's nothing yet I can do to help on this. I did rebuild, cherrypicking this [e3fe0dc89b1f](https://github.com/upx/upx/commit/e3fe0dc89b1f7b04135647ed5a0fc9e5b5e1d683)
Thanks for taking the time to track this down. I might learn a thing or two, so thanks for the details here and in the bug report.

**Comment by markus-oberhumer** [comment]
@jreiser I might have prematurely closed this issue - please re-open if this is not fixed yet.

**Comment by jreiser** [comment]
@romeojulietthotel If the program was a -fPIE executable, then commit d688a05ac78517bcba09bae0f60bc76f3aa51ddb fixes it for plain execve().  Some recent version of binutils (2.26.1 ?) no longer uses DT_JMPREL, so upx searches all DT_RELA for __libc_start_main.  Explicitly using ld-linux as a helper on the command line still draws a complaint " object file has no dynamic section" because ld-linux thinks it is loading a shared library.  Indeed, the file is ET_DYN; but running it as a main program should not care about any Elf64_Shdr.  Part of the problem is that a -fPIE main program has an identical file format to -fPIC shared library; only the usage intent differs (execve() main program vs DT_NEEDED shared library.)

**Comment by romeojulietthotel** [comment]
Here's what I did:

- cherrypicked [d688a05ac785](https://github.com/upx/upx/commit/d688a05ac78517bcba09bae0f60bc76f3aa51ddb)
- rebuilt upx
- added -pie and -fPIE to bash Makefile and rebuilt bash
- upx'd unstripped bash
- made the new upx'd bash my login shell
- rebuilt ffmpeg and observed no problems

Bash does not build -fPIE by default so I added that myself. I was not building as -fPIE before and I know no reason not to build it as -fPIE. It is strange because I came across an ubuntu bug that claimed the -fPIE compiled bash caused the errors that I reported here with xmalloc.

Just FYI, I will not test this with a non -fPIE bash.
